### ISSUE TITLE
Is there an open source "protobuf message validator" program?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4runtime#190]
Suppose I have a P4Info file contents that I have edited by hand, or generated from some program, e.g. p4c, and I want to check whether it has a "valid schema" that conforms to what is defined in the .proto files in this p4runtime repository.

Is there some existing open source program I can run, providing it the name of the P4Info file (preferably with a choice between any of JSON, text, or binary format, but at least one of those is better than none), plus the name of one or more of the p4runtime .proto files, and reports errors or "all ok"?

**Comment by antoninbas** [comment]
I believe you can use protoc with the `--encode` / `--decode` options. There are probably some third-party tools available online as well, or some third-party protoc plugins.
However, if you are specifically interested in validating P4Info messages, you can use the `pi_convert_p4info` binary installed by p4lang/PI. It is meant for conversion between different formats. However, it does validation implicitly when loading the P4Info message, and you can choose the same output format as input format... Feel free to submit patches to `pi_convert_p4info` as well if the tool is useful to you.

**Comment by jafingerhut** [comment]
OK, I tried out this command:
```
$ pi_convert_p4info -f prototext -t prototext -i $j -o $j.converted.txt
```
where `j` is a shell variable ranging over all of the files with names ending with the suffix ".p4info.txt" in this directory: https://github.com/p4lang/p4c/tree/master/testdata/p4_16_samples_outputs

Does the format `prototext` seem like the right one for those files?  I do not know the correspondence between the four file format names supported by the `pi_convert_p4info` command vs. the 3 formats that can be generated by `p4c`, but `prototext` gave the fewest error messages :-)

The command succeeded for all of the files except for one (more on that file below).

For the majority of files on which it succeeded, it seems like the `pkg_info` message near the beginning of those files was removed in the 'converted' output files:
```
pkg_info {
  arch: "psa"
}
```
Also, even though the output file is specified on the command line via `-o` (and the command requires this to be specified or it gives an error), it also prints the output to stdout as well.  Is that intentional?  Not a big deal, but seems a bit odd for a command where the output file must be given on the command line.

The one file where the command failed was from this command run in that directory:
```
$ pi_convert_p4info -f prototext -t prototext -i p4rt_digest_complex.p4.p4info.txt -o p4rt_digest_complex.p4.p4info.txt.converted.txt
Packed type for digest too complex
Error while importing protobuf message to p4info.
```
for the p4c repository as of this commit: https://github.com/p4lang/p4c/commit/b79021f377e5e6689c30a697c12bdd55b0d8d713

I do not understand why there is an error for that file.  Do you see anything wrong with it?

**Comment by jafingerhut** [comment]
Also, do you know if this is the correct syntax for a `prototext` format file that has a `map` value like those inside of the `P4TypeInfo` message, that has _multiple_ different key/value pairs?  In particular, is it supposed to be that the `new_types { ... }` is repeated for each key/value pair, if there is more than one?

https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/expected-p4info-files/psa-example-digest-bmv2.psa.p4info.txt#L85-L103

**Comment by antoninbas** [comment]
Yes in Protobuf text format you repeat the map name for each entry, just like for a repeated field.

Regarding the tool, I was a bit optimistic, it won't be able to do just validation of the Protobuf message. It does validation of the Protobuf message (which just means parsing the file and instantiating the in-memory Protobuf message), but it also transforms the P4Info message into an internal data structure for the PI library. This means that a lot of semantic validations are performed on the message (e.g. are the ids consistent?) but it also means that the P4Info will be rejected if the program is too complex for PI to handle (this must be what's happening with p4rt_digest_complex.p4). The internal representation is then converted to the desired format. Unfortunately in your case the whole process loses information: what is not relevant to PI is discarded (e.g. pkg info) and some structures are simplified / flattened.

I could modify the tool to do validation only. Note that the validation part itself is 3 lines of code (parsing of the file). If you are interested I suggest you write a Python script to do that instead. You can also make the script much better by adding semantic checks. You can check for consistent ids, you can check that the `P4TypeInfo` message is correct, ... It's a bit sad because `pi_convert_p4info` already does all of these things but unfortunately in the case of `pi_convert_p4info`, P4Info messages which are too complex for PI are also rejected and there is no way to separate the 2 checks.

I have been meaning to create a p4runtime-tools repository in the p4lang Github organization. It would be a good place for such a validation tool. It would also be a good place for a P4Runtime CLI eventually.

**Comment by jafingerhut** [comment]
I have hacked up this small Python program that may be a start at what you described, Antonin:

https://github.com/jafingerhut/p4-guide/blob/master/bin/parse-p4info.py

Current limitations:

+ It only supports reading the text format of P4Info, not the binary or JSON formats
+ It does not do any sanity checking, other than the syntax checking done by the Python library code that reads it, e.g. it checks that all field names are legal ones according to the '.proto' file message definitions, and that their corresponding values have the proper type (e.g. integer, string, LPM/EXACT/TERNARY keyword thing, etc.)
+ It does not write out a P4Info file after reading it, only a few summary statistics about the objects and type_info read.
+ It may use non-recommended ways of accessing the Python data structures created -- I did not know what the recommended methods were.  I just started hunting and peeking around in an interactive Python session trying to figure out which methods were available.  If there is some documentation you are aware of that describes better the methods available, I'd be happy to link to them from inside that program, for anyone else that wants to extend it.
